File: Data Availability/R Desurveying Codes/Average_Angle.R
#Average Angle Method

library(plot3D)
library(plot3Drgl)

Survey <- read.csv("#File#.csv", skip = 1, header = TRUE,)[,1:3] #Import drillhole survey data file

Dn <- #Input distance of interest here

Depth <- Survey[,"Depth..m."] #Call Depth data
AzimuthDegrees <- Survey[,"Az.Deg"] #Call Azimuth data
PlungeDegrees <- Survey[,"P.Deg"] #Call Plunge data

DepthDn <- append(Depth[Depth < Dn], Dn) #This step takes only Depth values < Dn value, adding Dn to the list

DepthDiffs <- diff(DepthDn)  #Distance along drillhole between survey points
AzimuthRadians <- (AzimuthDegrees * pi/180)  #Convert Azimuth degrees to radians
PlungeRadians <- (PlungeDegrees * pi/180)  #Convert Plunge degrees to radians

    #Difference between the total length intervals in all drillhole and distances up to Dn point of interest
ListLength <- (length(Depth) - length(DepthDiffs))
L <- length(Depth) - ListLength

D1 <- Depth[1:L] #Depth at first survey in each survey point pair
Az1 <- AzimuthRadians[1:L] #Azimuth at first survey in each survey point pair
P1 <- PlungeRadians[1:L] #Plunge at first survey in each survey point pair

D2 <- Depth[2:(L+1)] #Depth at second survey in each survey point pair
Az2 <- AzimuthRadians[2:(L+1)] #Azimuth at second survey in each survey point pair
P2 <- PlungeRadians[2:(L+1)] #Plunge at second survey in each survey point pair

Azimuth = (180/pi) * ((Az1[L] + Az2[L]) / 2) #Calculate Azimuth at Dn location
Plunge = (180/pi) * ((P1[L] + P2[L]) / 2) #Calculate Plunge at Dn location

Depth <- DepthDiffs*sin((P1 + P2)/2) #Difference in depths between survey points
North <- DepthDiffs*cos((P1 + P2)/2)*cos((Az1 + Az2)/2) #Difference in northing between survey points
East  <- DepthDiffs*cos((P1 + P2)/2)*sin((Az1 + Az2)/2) #Difference in easting between survey points

SumDepth <- round(c(0, cumsum(Depth)), digits = 2) #Sums the depths and adds collar coordinate
SumNorth <- round(c(0, cumsum(North)), digits = 2) #Sums the northings and adds collar coordinate
SumEast <- round(c(0, cumsum(East)), digits = 2) #Sums the eastings and adds collar coordinate

    #Plot 3D graph
scatter3Drgl(SumEast,
             SumNorth,
             SumDepth,
             colvar = NULL,
             type = "o",
             #xlim = c(#,#),
             #ylim = c(#,#),
             #zlim = c(#,#),
             xlab = "East (m)",
             ylab = "North (m)",
             zlab = "Depth (m)",
             ticktype = "detailed")

scatter3D(SumEast,
          SumNorth,
          SumDepth,
          colvar = NULL,
          type = "o",
          #xlim = c(#,#),
          #ylim = c(#,#),
          #zlim = c(#,#),
          xlab = "East (m)",
          ylab = "North (m)",
          zlab = "Depth (m)",
          phi = 30,
          theta = 30,
          ticktype = "detailed",
          nticks = 14,
          bty = "g")

    #Show on graph the East, North, Depth, Azimuth and Plunge at Dn location
legend("bottomleft",
       inset = c(0,0),
       legend = c(parse(text = sprintf('paste(East~(m),\': %s\')', SumEast[L+1])),
                  parse(text = sprintf('paste(North~(m),\': %s\')', SumNorth[L+1])),
                  parse(text = sprintf('paste(Depth~(m),\': %s\')', SumDepth[L+1])),
                  parse(text = sprintf('paste(Azimuth~(degree),\': %s\')', Azimuth)),
                  parse(text = sprintf('paste(Plunge~(degree),\': %s\')', Plunge))),
       bty = "n",
       cex = 0.6)

File: Data Availability/Python Desurveying Codes/Average_Angle.py
# -*- coding: utf-8 -*-
"""
Created: 26 May 2020

@author: Benjamin Williams
"""

import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

Survey = np.genfromtxt('#File#.csv', delimiter = ',', skip_header=2, usecols=(0,1,2)) #Import drillhole survey data file

Dn = #Input distance of interest here

Depth = Survey[:, 0] #Call Depth data
AzimuthDegrees = Survey[:, 1] #Call Azimuth data
PlungeDegrees  = Survey[:, 2] #Call Plunge data

DepthDn = np.append(Depth[Depth < Dn], Dn)    #This step takes only Depth values < Dn value, adding Dn to the list

DepthDiffs = np.diff(DepthDn)  #Distance along drillhole between survey points
AzimuthRadians = np.radians(AzimuthDegrees) #Convert Azimuth degrees to radians
PlungeRadians = np.radians(PlungeDegrees) #Convert Plunge degrees to radians

ListLength = (len(Depth) - len(DepthDiffs)) * -1  #Difference between the total length intervals in all drillhole and distances up to Dn point of interest

Az1  = AzimuthRadians[:ListLength] #Azimuth at first survey in each survey point pair - radians
P1   = PlungeRadians[:ListLength] #Plunge at first survey in each survey point pair - radians
Az1D = AzimuthDegrees[:ListLength] #Azimuth at first survey in each survey point pair - degrees
P1D  = PlungeDegrees[:ListLength] #Plunge at first survey in each survey point pair - degrees

    #If statement for if list = 0, i.e. Dn > penultimate survey point depth
if ListLength + 1 < 0:
    ListLength = ListLength + 1
else:
    ListLength = None

Az2  = AzimuthRadians[1:ListLength] #Azimuth at second survey in each survey point pair - radians
P2   = PlungeRadians[1:ListLength] #Plunge at second survey in each survey point pair - radians
Az2D = AzimuthDegrees[1:ListLength] #Azimuth at second survey in each survey point pair - degrees
P2D  = PlungeDegrees[1:ListLength] #Plunge at second survey in each survey point pair - degrees

    #Calculate Azimuth and Plunge for Dn location
Azimuth = (Az1D + Az2D) / 2
Plunge = (P1D + P2D) / 2

    #Calculate and sum all depths, northings and eastings cumulatively down the drillhole
Depth = DepthDiffs*np.sin((P1 + P2)/2)
North = DepthDiffs*np.cos((P1 + P2)/2)*np.cos((Az1 + Az2)/2)
East  = DepthDiffs*np.cos((P1 + P2)/2)*np.sin((Az1 + Az2)/2)

SumDepth = np.cumsum(Depth)
SumNorth = np.cumsum(North)
SumEast  = np.cumsum(East)

SumDepth = np.insert(SumDepth, 0, 0.0)
SumNorth = np.insert(SumNorth, 0, 0.0)
SumEast  = np.insert(SumEast,  0, 0.0)

    #Plot 3D graph
print('Depth at point:', SumDepth[-1])
print('North at point:', SumNorth[-1])
print('East at point:', SumEast[-1])
print('Azimuth at Dn Depth:', Azimuth[-1])
print('Plunge at Dn Depth:', Plunge[-1])

ax = plt.axes(projection='3d')

ax.scatter3D(SumEast, SumNorth, SumDepth, color='blue')
ax.plot3D(SumEast, SumNorth, SumDepth, 'blue')

#ax.set_xlim(#, #)
#ax.set_ylim(#, #)
#ax.set_zlim(#, #)
ax.set_xlabel('East (m)')
ax.set_ylabel('North (m)')
ax.set_zlabel('Depth (m)')

ax.set_title('Average Angle Method')

plt.show()